### ISSUE TITLE
Filter UI should update immediately even if data is loading

### ISSUE THREAD

**Issue by Eric-Arellano** [ParkingReformNetwork/reform-map#593]
While showing the map to someone with bad data, they were confused when clicking the toggle for only places with parking minimums repealed because they clicked the toggle but nothing happened for half a second. So they tried to click it again. 

The UI should update immediately, even if the data is still loading.

--

Update: This applies to all the filter options. It is particularly egregious when lots of places need to be rendered, like when pressing the toggle to go from the initial view with only 96 places to all 3200 places. Post-https://github.com/ParkingReformNetwork/reform-map/pull/602, I suspect the issue is DOM rendering.

**Comment by Eric-Arellano** [comment]
I'm making good progress on this:

* https://github.com/ParkingReformNetwork/reform-map/pull/602
* https://github.com/ParkingReformNetwork/reform-map/pull/603

Benchmarks in https://github.com/ParkingReformNetwork/reform-map/pull/604 show two possible optimizations:

* Don't update the table when in map view and vice versa. These are the slowest operations
* Fix that the scorecard initialization double triggers running the filter manager on its initial run

https://github.com/ParkingReformNetwork/reform-map/blob/1cbfc3706c5a1ecc88c432e98d8973d7cbe34f63/src/js/scorecard.ts#L129-L132

**Comment by Eric-Arellano** [comment]
Claude Opus has good suggestions.

The core issue
Every filter change triggers Observable.notify(), which synchronously runs all subscribers (markers, table, counters, sliders). Two things make that expensive at 6038 places, and both are cheaper/safer to fix than the setTimeout idea you're worried about.

Tier 1 — cheap, safe, high impact
1. placeIds rebuilds a 6038-element Set on every access

[FilterState.ts:147-149](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/state/FilterState.ts#L147-L149) does new Set(Object.keys(this.matchedPlaces)) fresh each call. Per single filter toggle it's hit ~4×: [counters.ts:277](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/filter-features/counters.ts#L277) + [:286](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/filter-features/counters.ts#L286) (twice, just to read .size), and [markers.ts:90,94](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/map-features/markers.ts#L90) (twice). So you allocate ~4 six-thousand-element Sets per click, and the counters ones only want the count.

Fix: memoize the Set inside CacheEntry (compute once per state change), and add a get numMatchedPlaces() returning Object.keys(matchedPlaces).length so counters never build a Set at all. Purely mechanical, no behavior change.

2. Every cache getter re-runs a deep isEqual over the whole FilterState

[FilterState.ts:189](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/state/FilterState.ts#L189) — ensureCache() does isEqual(currentState, this.cache.state) (lodash deep-compare over several Sets) on every access to placeIds/matchedCountries/matchedPlaceTypes/matchedPolicyTypes/matchedPlaces. That's ~8–10 deep comparisons per toggle. Replace with a monotonic version counter bumped in update() and compared as an integer. Trivial and removes lodash isEqual from the hot path.

3. zoomend recomputes determineIsPrimary for all 6038 markers

[markers.ts:111-120](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/map-features/markers.ts#L111-L120) recomputes primacy (which filters benefit_district arrays) for every marker on every zoom tick. You already compute isPrimary at marker creation ([markers.ts:59](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/map-features/markers.ts#L59)) — stash it on the marker and read it back instead of recomputing.

Tier 2 — the deferral idea, done without the bug risk
4. Coalesce updates with requestAnimationFrame, not setTimeout

Your instinct that a raw setTimeout is risky is right — a naive setTimeout chain can fire with stale state, double-apply, or race the view toggle. The robust version of the same idea is a rAF-coalesced scheduler: on state change, cancelAnimationFrame(pending) then schedule one frame that reads the latest state and updates the active view. Because it always cancels the prior frame and reads current state at fire time, rapid changes collapse to one render and there's no staleness. This is what actually delivers "the checkbox/slider repaints instantly, heavy work happens next frame" — the exact DOM-render delay called out in #593.

5. The population slider is the worst offender and isn't throttled at all

[populationSlider.ts:159-165](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/filter-features/populationSlider.ts#L159-L165) calls filterManager.update() on every input event — i.e. every pixel of drag fires a full O(6038) recompute + Leaflet marker diff + counter innerHTML rewrite. If you add the rAF scheduler from #4, dragging the slider automatically collapses to one render per frame. This single change is probably the biggest "feels janky" win, and it's contained to the slider.

Tier 3 — load time / bundle (2.08MB expand)
6. Parse the data via JSON.parse, not as a JS module. [data.ts:126](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/model/data.ts#L126) imports core.json as a module, which Parcel may inline as a JS object literal. V8 parses JSON.parse("…") of a big string ~1.5–2× faster than an equivalent object literal, and it lets the 3MB download start in parallel and be served pre-compressed. Worth a quick check of how Parcel emits it, then switching to fetch() + JSON.parse.

7. readData eagerly deep-copies all 6038 entries + parses every Date on startup ([data.ts:104-135](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/model/data.ts#L104-L135)). The {...raw, url} spread and Date.fromNullable on every policy run before first paint. Computing url lazily and parsing dates on demand (only the table sorts by them) trims main-thread blocking at load.

8. FontAwesome runtime is your heaviest non-data dependency. You already tree-shake to ~15 icons ([fontAwesome.ts](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/layout/fontAwesome.ts)), but dom.watch() + SVG core ships a live DOM-mutation observer. Inlining those 15 SVGs and dropping @fortawesome/* entirely is the single biggest bundle cut available — worth measuring against effort.

My recommendation: do Tier 1 (#1–3) + the rAF scheduler (#4–5) as one focused PR — all low-risk, and together they address the toggle/drag jank far more safely than a bare setTimeout. Tier 3 is a separate load-time investigation.

**Comment by Eric-Arellano** [comment]
I think this is done for now. Feedback on Claude's suggestions:

- ideas 1-3 were great and done in https://github.com/ParkingReformNetwork/reform-map/pull/925, https://github.com/ParkingReformNetwork/reform-map/pull/926, and https://github.com/ParkingReformNetwork/reform-map/pull/928
- FontAwesome (idea 8) was a huge deal https://github.com/ParkingReformNetwork/reform-map/pull/931
- Idea 6 about JSON.parse is a false promise. Parcel already uses JSON.parse
- Idea 7 is also a false promise. We need the Date eagerly. The only thing we could make lazy is the URL, but that's not gonna buy much

That leaves ideas 4-5 to use requestAnimationFrame. I'd rather not do that unless we _really_ need to because it adds a lot of complexity. The FontAwesome speedup was so substantial that I don't think we need it.

**Comment by Eric-Arellano** [comment]
Actually, there is more to do. 

<details><summary>New Claude Code analysis</summary>

I temporarily instrumented [main.ts](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/main.ts) with `performance.now()` around each init phase, built, and measured 4 cold loads via Playwright. The `initApp()` synchronous window (which is the entire ~210ms gap between "data fetched" ≈260ms and "counter ready" ≈470ms in [no-fontawesome.json](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/benchmark-results/no-fontawesome.json)) breaks down as:

| Phase | median ms | % |
| --- | --- | --- |
| initSearch (Choices.js) | 93.5 | 41% |
| initialize / notify | 58.7 | 26% |
| readData | 46.1 | 20% |
| initPlaceMarkers | 10.9 | 5% |
| preData (map, popups) | 10.6 | 5% |
| initTable | 6.2 | 3% |
| filterOptions+share, scorecard, counters | ~3 | 1% |
| TOTAL | ~229 | |

The key structural fact: **all 229ms is synchronous and runs before the browser can paint the map.** There's no `await` that yields between `readData()` and the end, so markers can't show until every one of these finishes — including the search box, which the map's own code comment calls a "power user" feature.

Companion micro-benchmark (Node, this machine): `JSON.parse` ~7.6ms, Luxon parsing all 949 dates ~4ms, full data processing ~7ms (2ms without Luxon).

## Recommendations, in priority order

### 1. Lazily initialize Choices.js search — biggest single win (~93ms)

`new Choices(...)` with 6,069 items in [search.ts:61](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/search.ts#L61) costs 93ms and blocks the map paint. Search is opened by clicking the search icon ([search.ts:25](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/search.ts#L25)), and it's explicitly a power-user feature. Defer construction until the popup is first opened (or until an idle callback after first paint). This removes ~40% of the load JS from the critical path. You'll need to route the `filterManager.subscribe("reset search UI…")` and the `change` listener through a "once initialized" guard, but it's mechanical.

### 2. Yield to the browser after markers are inserted, then build deferred UI

Even after fixing search, `initSearch`/`initTable`/`initScorecard` don't need to run before the map is visible. Restructure `initApp` so the critical path is: `readData → filterManager → initPlaceMarkers → initCounters → filterManager.initialize()` (markers paint), then `await` a macrotask / `requestIdleCallback` and build search, table, and scorecard. This lets the map paint ~80–130ms sooner in felt load. The table already lazily refreshes on view-toggle, so it's safe to build late.

### 3. `initialize / notify` = 59ms is mostly SVG marker insertion — try the Canvas renderer

`filterManager.initialize()` inserts ~6,000 `CircleMarker`s, each an SVG `<path>` DOM node ([markers.ts:46-50](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/map-features/markers.ts#L46-L50)). That DOM churn is the bulk of the 59ms and also drives the post-JS paint. Leaflet's Canvas renderer (`new Map(..., { preferCanvas: true })` or a shared `L.canvas()`) draws all circles into one canvas — far cheaper to insert and repaint, and it also speeds up the filter hot path (see below). Tooltips still work on canvas CircleMarkers. This is the one change that needs careful manual testing (hover tooltips, the `NO_MANDATES_MARKERS_PANE` z-index workaround, zoom `setRadius`), so I'd gate it behind a benchmark.

### 4. Parse dates lazily (small, easy: ~4ms load + mobile multiplier)

`readData` eagerly builds a Luxon `DateTime` for all 949 dates via [types.ts:8-14](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/model/types.ts#L8). But the only load-time consumer is the year filter, which does `date.parsed.year.toString()` ([FilterState.ts:269](vscode-webview://1qlnrcvn5127evgjr1g0j4l32g81mvd45729s0eoqvniohojnips/src/js/state/FilterState.ts#L269)) — the year is just `raw.slice(0,4)`. `format()` and sort comparisons only matter once the table is open. Make `parsed` a lazy getter (or store `raw` and parse on demand). Luxon is ~4ms here but multiplies 4–6× on mid-tier mobile, and it also cuts allocations during the 46ms `readData`.

### Filter hot path note

The filter benchmark shows "any reform" reset at 27ms JS vs "reduce min" at 11ms — the asymmetry is because "any reform" matches ~6,000 places and `ensureCache` recomputes everything plus `updatePlaceVisibility` adds thousands of SVG nodes. The Canvas renderer (#3) would help here too. This path is already reasonable, so I'd treat it as lower priority than initial load.

</details>

**Comment by Eric-Arellano** [comment]
This is now complete.

```
Metric                          Before         After         Delta   Delta %
----------------------------------------------------------------------------
initialLoad                    2727 ms        316 ms      -2411 ms    -88.4%
  network                         1 ms          1 ms         -0 ms     -7.7%
  firstPaint                    119 ms        115 ms         -4 ms     -3.3%
  dataFetch                      75 ms         76 ms         +1 ms     +1.2%
  jsBuild                      2510 ms         72 ms      -2438 ms    -97.1%
  markerPaint                    23 ms         28 ms         +5 ms    +24.0%
tableView                       155 ms        143 ms        -12 ms     -7.7%
filterReduceMin                  17 ms         21 ms         +4 ms    +24.9%
  js                             12 ms         10 ms         -2 ms    -13.4%
  paint                           5 ms         11 ms         +6 ms   +114.0%
filterReset                    1252 ms         25 ms      -1227 ms    -98.0%
  js                             27 ms         10 ms        -17 ms    -62.4%
  paint                        1225 ms         15 ms      -1210 ms    -98.8%
loadSearch                       48 ms        134 ms        +85 ms   +175.3%
  js                              0 ms         89 ms        +88 ms +44225.0%
  paint                          48 ms         45 ms         -3 ms     -7.2%
transfer                       3.03 MB       2.89 MB      -0.14 MB     -4.8%
```
